### ISSUE TITLE
Negotiate PSKs

### ISSUE THREAD

**Issue by bifurcation** [mlswg/mls-protocol#367]
[This issue is part of a feature breakdown based on #336]

There are a few subissues here:

# 1. Signaling which PSK(s) to add to the key schedule

When a PSK is to be added to the key schedule in an epoch, the sender of the Commit needs to signal this to existing members in a Commit, and to any new members in the GroupInfo.  In both cases, there should be an extension to signal this.  This would work like the [TLS ServerHello PSK extension](https://tools.ietf.org/html/rfc8446#section-4.2.11), which simply indicates which PSK is being used.  In the MLS case, we would probably include the full ID instead of just an index, since there's not a clear list like in TLS.  And just like the server's ability to complete the TLS handshake proves that it has the PSK, the confirmation value in MLS would validate the same.

```
struct {
    opaque identity<1..2^16-1>;
} PreSharedKeyExtension;
```

# 2. Signaling which PSKs are available

It might also be useful for clients to advertise in a KeyPackage extension what PSKs they have available.  This would be analogous to the ClientHello PSK extension, which includes a binding value to prove that the client has the PSK.  If we define such an extension, we will need to decide (a) whether we need such a binding and (b) what parts of the KeyPackage should be covered by the binding MAC.

Unlike TLS, however, the set of available PSKs might change over the lifetime of the group.  Since the KeyPackages in the leaves of the tree can be modified, in principle, clients could signal the availability of new PSKs by updating, and you would get a clear assurance that everyone has the key before you try to use it.  But this seems pretty complex, and for not much benefit if the application can coordinate which PSKs are available.

# 3. Structuring PSK input / multiple PSKs

It may be desirable for the input to the key schedule not to be a PSK directly, but some derived value that adds some context to the PSK.  This is the approach taken in the [TLS external PSK draft](https://tools.ietf.org/html/draft-ietf-tls-external-psk-importer-05) and in #336.  We might also want to support the application of multiple PSKs in a single epoch.

For the former need, we would need to define what context is added, how it is signaled in the relevant extensions, and how it is combined with the PSK itself to derive the ultimate PSK.  For the latter need, we need a way to signal multiple PSKs and their order, and an algorithm for combining them all into one value to be added to the key schedule.  (The schemes in both #336 and #337 are relevant here.)

**Comment by kkohbrok** [comment]
Except for the advertisement of PSKs in KeyPackages, we already covered all of this in #336. In particular, it is important to distinguish PSKs by type (internal/external, which type of internal) by having an extra layer of derivations before injecting the PSK into the key schedule. Again, this is already present in #336.

PR #376 re-implements #336 partially, but instead of having additional types of proposals, it instead has a dedicated preSharedKey extension that can come with a commit. Why an extension instead of a proposal? I would think that PSKs are part of MLS' core functionality, for example to securely create subgroups or to restart groups with different parameters.

Having PSKs in KeyPackages might be a good idea, but as you already hint at, I don't think it's a good idea to use it as a necessary way of signalling the availability for all PSKs. With group restarting and branching, we use PSKs much more dynamically than in TLS. For example, when creating a sub-group, the creator already knows that the relevant PSKs are available to all members due to their membership in the original group.

Do you want me to isolate the parts of #336 relevant for this particular issue and create a separate PR? If we still want an extension instead of a dedicated proposal, we should probably merge #369 first and then do the PR against the result.

**Comment by bifurcation** [comment]
Fixed by #336
